proj_eclipse/src/main/java/smartmag/models/BoxModel.java

```diff
@@ -0,0 +1,245 @@
+				return instances.get(b.getIndirizzo());
+			}
+		} else
+			throw new IllegalArgumentException("box non valido");
+	}
+
+	/**
+	 * metodo costruttore della classe BoxModel
+	 * 
+	 * @param b box del quale si vuole generare il modello
+	 */
+	private BoxModel(Box b) {
+		this.record = fetchBoxByIndirizzo(b.getIndirizzo());
+		this.box = b;
+	}
+
+	/**
+	 * controlla che ci sia un record del box nel db
+	 * 
+	 * @return true se é presente, false se non lo é
+	 */
+	public boolean isSavedInDb() {
+		refreshFromDb();
+		return record != null;
+	}
+
+	/**
+	 * restituisce l'oggetto Box rappresentato dal modello
+	 * 
+	 * @return
+	 */
+	protected Box getBox() {
+		return box;
+	}
+
+	protected void setBox(Box b) {
+		if (b != null && b.isValid() && (DSL.select().from(PRODOTTO)
+				.where(PRODOTTO.ID.eq(b.getProd().getId())) != null)) {
+			this.box = b;
+		} else
+			throw new IllegalArgumentException("box o prodotto non valido");
+
+	}
+
+	public BoxRecord getRecord() {
+		return record;
+	}
+
+	/**
+	 * permette di creare il record del box nel database, solo se non é giá
+	 * presente
+	 * 
+	 * @throws SQLIntegrityConstraintViolationException se il record é giá
+	 *                                                  presente nel db
+	 */
+	public void createBox() throws SQLIntegrityConstraintViolationException {
+		if (isSavedInDb()) {
+			throw new SQLIntegrityConstraintViolationException(
+					"il box#" + box.getIndirizzo() + " esiste giá");
+		}
+		BoxRecord r = DSL.newRecord(BOX);
+		r.setId(box.getIndirizzo());
+		copyBoxIntoRecord(box, r);
+		r.store();
+		this.record = r;
+	}
+
+	/**
+	 * aggiorna i dati del record e del box del modello usando quelli presenti
+	 * nel db
+	 */
+	private void refreshFromDb() {
+		this.record = fetchBoxByIndirizzo(box.getIndirizzo());
+		if (this.record != null) {
+			Box b = boxFromRecord(this.record);
+			this.box.setIndirizzo(b.getIndirizzo());
+			this.box.setQuantità(b.getQuantità());
+			this.box.setProd(b.getProd());
+		}
+
+		// TODO: event
+	}
+
+	/**
+	 * permette di cambiare il prodotto di un box solo se il prodotto giá
+	 * presente nel box é effettivamente finito e il nuovo prodotto esiste nel
+	 * db
+	 * 
+	 * @param p nuovo prodotto da assegnare al box
+	 */
+	public void cambiaProdotto(Prodotto p) {
+
+		if (box != null && box.isValid() && box.getQuantità() == 0) {
+			if (p != null && p.isValid() && (DSL.select().from(PRODOTTO)
+					.where(PRODOTTO.ID.eq(p.getId()))) != null) {
+				box.setProd(p);
+			} else
+				throw new IllegalArgumentException("prodotto non valido");
+		} else
+			throw new IllegalArgumentException(
+					"attenzione: il box deve essere valido e vuoto");
+	}
+
+	/**
+	 * cancella il record dell'ordine
+	 */
+	public void deleteRecord() {
+		if (isSavedInDb()) {
+			record.delete();
+		}
+	}
+
+	/**
+	 * incrementa la quantita di prodotto in un box
+	 * 
+	 * @param quantita numero di unita di prodotto aggiunte
+	 */
+	public void rifornisci(int quantita) {
```

Comment:
Essendo un'operazione delicata la metterei _protected_ in modo che solo i modelli possono modificare la qta.
Principalmente sarà il modello delle Movimentazioni a farlo.